### ISSUE TITLE
Unpin tornado dependency

### ISSUE THREAD

**Issue by ian-r-rose** [jupyterlab/jupyterlab#6131]
#6123 rolled back our dependency on `notebook` due to some upstream issues with async kernel starts (or possibly tornado 6). We need to identify and fix these issues before making a release.

Edit: we have resolved the immediate issue with notebook in 5.7.8, but issues with tornado 6 remain.

**Comment by ian-r-rose** [comment]
cc @minrk Do you have any idea what might be going on here?

**Comment by minrk** [comment]
tornado 6 seems most likely to be the culprit, though if you have notebook 5.7.5 *and* tornado 6, you should be fine (at least all the notebook's own tests pass with tornado 6 and 5.7.5).

**Comment by ian-r-rose** [comment]
These problems arise when both notebook 5.7.5 and tornado 6 are installed. Individually they seem to cause problems as well, but different ones. I can post some representative errors later today, but my initial look seemed to indicate that the `maybe_future` changes seem to be causing some problems. If I switch back to using `gen.maybe_future` in `zmqhandlers` then things seem to work better.

**Comment by ian-r-rose** [comment]
Ah, I see you experimenting in #6138, thanks!

**Comment by minrk** [comment]
I found a regression in 5.7.5 in https://github.com/jupyter/notebook/pull/4513 that's causing the restart issues. The remaining tornado 6 failures I believe are caused by https://github.com/jupyter/terminado/issues/62 which ought to be fixed in terminado master

**Comment by ian-r-rose** [comment]
Awesome, thanks for diving into this @minrk. I was rooting around the notebook zmq code yesterday and not making much progress :)

**Comment by mathematicalmichael** [comment]
@minrk is this why I sometimes get really laggy restart kernels? also unable to download files via jupyterlab. using `docker-stacks/datascience-notebook`

**Comment by minrk** [comment]
@mathematicalmichael for restarts, I believe so. The download issue is fixed by #6139

**Comment by mathematicalmichael** [comment]
@minrk sorry if this is a dumb question, but if I purged my docker cache and re-pulled the latest `docker-stacks/scipy-notebook` but still cannot download and experience laggy restarts. Any guess as to when the relevant changes will get into that software stack? looks like terminado has been released, and stacks updated nearly daily, so I'm perplexed as to why the fixes haven't propagated yet.

**Comment by ian-r-rose** [comment]
A git bisect of tornado seems to indicate [this](https://github.com/tornadoweb/tornado/commit/a55c7fe7abcfd3e11cf47efb7759d680a2dac35d) commit as the source of the troubles we are having.

**Comment by ian-r-rose** [comment]
Reopening this, but with a new focus on just unpinning tornado so we can use tornado 6.

**Comment by Zsailer** [comment]
>A git bisect of tornado seems to indicate [this](https://github.com/tornadoweb/tornado/commit/a55c7fe7abcfd3e11cf47efb7759d680a2dac35d) commit as the source of the troubles we are having.

@ian-r-rose good work! I'm reviewing this commit in tornado now. There changes in `iostream.py` may be the cause for our troubles.

**Comment by Zsailer** [comment]
I can verify, [the commit](https://github.com/tornadoweb/tornado/commit/a55c7fe7abcfd3e11cf47efb7759d680a2dac35d) mentioned above causes the `"Already Reading"` error. I rolled back to the commit before, and all tests pass. Bump to this commit and get the error.

**Comment by jasongrout** [comment]
Just to make progress on this: do we think it is a tornado issue, or a notebook issue? If the former, has anyone opened a tornado github issue?

**Comment by Zsailer** [comment]
_Thanks to @ian-r-rose for tracking down the commit above._

I've explored this further. I *think* this is a bug in Tornado. 

I've tracked down the issue to a single change in Tornado that broke jupyterlab's tests. It looks like tornado removed a line in their `IOStream` class that checks that all `Future`s are done when handling an HTTP response. I'm not sure if they removed this line intentionally or accidentally. 

I can fix this issue by reverting to previous behavior. This ensures that all futures started by an HTTP request are done before closing its connection. See the diff for the fix [here](https://github.com/Zsailer/tornado/commit/60d67236e736973a96b9d628baa3bc6f628fa1b1). 

Before I open an issue on Tornado, I need to wrap my brain around why this breaks. It's also possible we're missing an `await` call somewhere in the notebook server. I'll update when I figure this out.

**Comment by jasongrout** [comment]
It sounds to me like you have enough understanding to at least open an issue in the tornado repo. If it was an unintentional removal, they may be able to see and fix it quickly.

**Comment by ian-r-rose** [comment]
Thanks @Zsailer for doing the hard detective work! I am also in favor of opening a tornado issue.

**Comment by Zsailer** [comment]
Ok, I've open an issue in the tornado repo. 🤞

**Comment by Zsailer** [comment]
Great news! 

Looks like this was a bug in Tornado after all. I'm submitting a PR today and should see the fix land in Tornado soon.

**Comment by saulshanabrook** [comment]
We are waiting for the Tornado PR to be merged, which is blocked on them adding a test for this behavior.

**Comment by Zsailer** [comment]
Quick update here:

We've been struggling to build a test in Tornado that triggers this issue. After trying for a long time, we decided to abandon writing a test and just let jupyterlab be an indirect test of the issue. 

I've got a new PR (#[2670](https://github.com/tornadoweb/tornado/pull/2670)) open in Tornado that will close this issue. I should be merged *very* soon. Sorry for the delay on this!

**Comment by ian-r-rose** [comment]
Thanks for the dogged work on this @Zsailer !

**Comment by ian-r-rose** [comment]
@Zsailer any updates on the Tornado PR?

**Comment by Zsailer** [comment]
Nothing yet. I'm still waiting for it to be merged. I just pinged the tornado team on the PR to see if they need anything from me.

**Comment by Zsailer** [comment]
Good news! tornadoweb/tornado#2670 was successfully merged. We should be able to unpin tornado moving forward.

**Comment by ian-r-rose** [comment]
Great! Now we just need to convince them to make a patch release ;)

**Comment by jasongrout** [comment]
It looks like they may be making a release this weekend: https://github.com/tornadoweb/tornado/pull/2670#issuecomment-504610363

**Comment by Zsailer** [comment]
We've got a Tornado patch release 🎉  6.0.3 is now out!

**Comment by vidartf** [comment]
I think it feels appropriate for you to do the PR that unpins the dependency @Zsailer 🥇

**Comment by Zsailer** [comment]
@vidartf Your wish is my command 😄
